Command Line Sensor - json_attributes

alexhardwicke opened the pull request:
## Description:
Add json_attributes support to the command line sensor (functions in the same way as for the rest component).

It's bothered me for a while that command line sensors can't have attributes, and as one already has access to value_json in command line, json_attributes didn't feel out of place.

Also, I'm completely new to python (or well, started six weeks ago) and thought this would be a simple feature to hopefully start contributing a bit more to home assistant.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5889

## Example entry for `configuration.yaml` (if applicable):
```yaml
  - platform: command_line
    command: 'python3 my_json_script.py'
    value_template: '{{ value_json.primary_value }}'
    json_attributes:
      - secondary_value
      - tertiary_value
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

Commit:
Add tests to command_line for json_attrs

Commit:
Add json_attrs to command_line

homeassistant commented:
Hi @alexhardwicke, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#15679).

Once you do that we will be able to review and accept this pull request.

Thanks!

Commit:
Remove whitespace on blank line

alexhardwicke commented on homeassistant/components/sensor/command_line.py:
```diff
@@ -81,11 +89,33 @@ def state(self):
+                    else:
+                        _LOGGER.warning("JSON result was not a dictionary")
+                except ValueError:
+                    _LOGGER.warning("Command result could not be parsed as JSON")
```
I can split this onto two lines if you'd like, but it's an exact copy from the rest sensor, so I wasn't sure if < 80 was a hard rule, or if it's just that rest snuck past (or pre-dates) the checks.

Commit:
Stick to <80 row length

awarecan commented on tests/components/sensor/test_command_line.py:
```diff
@@ -68,3 +73,56 @@ def test_bad_command(self):
+        """Test attributes get extracted from a JSON result."""
+        data = command_line.CommandSensorData(
+            self.hass,
+            'echo { \\"key\\": \\"some_json_value\\" }', 15
```
I would like to test more item in the JSON duct. Let's say you have 3 item in the output, and only pick up 2 item as json_attrs

alexhardwicke replied:
Done. Added two additional keys (and wanted both) to the original test, and added two more tests: One where the sensor config requests a key missing in the command's output, and one where the command's output has a key we don't request. I check in both cases that the key isn't in attributes.

awarecan commented on homeassistant/components/sensor/command_line.py:
```diff
@@ -81,11 +89,34 @@ def state(self):
+                    else:
+                        _LOGGER.warning("JSON result was not a dictionary")
+                except ValueError:
+                    _LOGGER.warning("Command result could not be parsed as" +
```
In Python, you don't need +

alexhardwicke replied:
Done, thanks - good to know.

awarecan commented on homeassistant/components/sensor/command_line.py:
```diff
@@ -81,11 +89,34 @@ def state(self):
+                try:
+                    json_dict = json.loads(value)
+                    if isinstance(json_dict, dict):
+                        attrs = {k: json_dict[k] for k in self._json_attrs
```
You can save `attrs` var

awarecan commented on homeassistant/components/sensor/command_line.py:
```diff
@@ -81,11 +89,34 @@ def state(self):
+            if value:
+                try:
+                    json_dict = json.loads(value)
+                    if isinstance(json_dict, dict):
```
Change `dict` to `collections.Mapping`

awarecan commented on homeassistant/components/sensor/command_line.py:
```diff
@@ -27,6 +28,7 @@
 
 SCAN_INTERVAL = timedelta(seconds=60)
 
+CONF_JSON_ATTRS = 'json_attributes'
```
I would use full name CONF_JSON_ATTRIBUTES

alexhardwicke replied:
Done, have renamed a few other variables using attrs to attributes at the same time.

awarecan commented on homeassistant/components/sensor/command_line.py:
```diff
@@ -35,6 +37,7 @@
     vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
     vol.Optional(CONF_UNIT_OF_MEASUREMENT): cv.string,
     vol.Optional(CONF_VALUE_TEMPLATE): cv.template,
+    vol.Optional(CONF_JSON_ATTRS, default=[]): cv.ensure_list_csv,
```
No need give default

awarecan reviewed (commented):
Good job for first time contributor.

Commit:
Use collections.Mapping, not dict

Commit:
Rename *attrs to *attributes

Commit:
Remove extraneous + for string concat

Commit:
Test multiple keys

Commit:
Add test

Makes sure the sensor's attributes don't contain a value for a missing key,
even if we want that key.

Commit:
Test that unwanted keys are skipped

MartinHjelmare commented on homeassistant/components/sensor/command_line.py:
```diff
@@ -81,11 +90,34 @@ def state(self):
+                except ValueError:
+                    _LOGGER.warning("Command result could not be parsed as"
+                                    "JSON")
+                    _LOGGER.debug("Erroneous JSON: %s", value)
```
Don't double log. Move the value to the warning.

alexhardwicke replied:
Right-o, done, thanks. :)

Commit:
Remove additional log line

Commit:
Update tests for log changes

fabaff committed:
Fix ordering

fabaff reviewed (approved):
Thanks :bird:
